typing/typedecl.ml

```diff
@@ -2066,27 +2066,57 @@ let report_error ppf = function
-             "In this definition, a type variable has a variance that"
-             "is not reflected by its occurrence in type parameters."
-       | No_variable ->
-           fprintf ppf "@[%s@ %s@]"
-             "In this definition, a type variable cannot be deduced"
-             "from the type parameters."
-       | Variance_not_deducible ->
-           fprintf ppf "@[%s@ %s@ It"
-             "In this definition, a type variable has a variance that"
-             "cannot be deduced from the type parameters."
+       | Variance_variable_error { error; variable; context } ->
+           Printtyp.prepare_for_printing [ variable ];
+           begin match context with
+           | Type_declaration (id, decl) ->
+               Printtyp.add_type_declaration_to_preparation id decl;
+               fprintf ppf "@[<v>%s@;<1 2>%a@;"
+                 "In the definition"
+                 (Printtyp.prepared_type_declaration id)
+                 decl
+           | Gadt_constructor c ->
+               Printtyp.add_constructor_to_preparation c;
+               fprintf ppf "@[<v>%s@;<1 2>%a@;"
+                 "In the GADT constructor"
+                 Printtyp.prepared_constructor
+                 c
+           | Extension_constructor (id, e) ->
+               Printtyp.add_extension_constructor_to_preparation e;
+               fprintf ppf "@[<v>%s@;<1 2>%a@;"
+                 "In the extension constructor"
+                 (Printtyp.prepared_extension_constructor id)
+                 e
+           end;
+           begin match error with
+           | Variance_not_reflected ->
+               fprintf ppf "@[%s@ %a@ %s@ %s@ It"
+                 "the type variable"
+                 Printtyp.prepared_type_expr variable
+                 "has a variance that"
+                 "is not reflected by its occurrence in type parameters."
+           | No_variable ->
+               fprintf ppf "@[%s@ %a@ %s@ %s@]@]"
+                 "the type variable"
+                 Printtyp.prepared_type_expr variable
+                 "cannot be deduced"
+                 "from the type parameters."
+           | Variance_not_deducible ->
+               fprintf ppf "@[%s@ %a@ %s@ %s@ It"
+                 "the type variable"
+                 Printtyp.prepared_type_expr variable
+                 "has a variance that"
+                 "cannot be deduced from the type parameters."
+           end
        | Variance_not_satisfied n ->
-           fprintf ppf "@[%s@ %s@ The %d%s type parameter"
+           fprintf ppf "@[@[%s@ %s@ The %d%s type parameter"
              "In this definition, expected parameter"
              "variances are not satisfied."
              n (Misc.ordinal_suffix n));
       (match n with
-       | No_variable -> ()
+       | Variance_variable_error { error = No_variable; _ } -> ()
        | _ ->
-           fprintf ppf " was expected to be %s,@ but it is %s.@]"
+           fprintf ppf " was expected to be %s,@ but it is %s.@]@]"
```

Comment:
This `Variance Typedecl.Bad_variance` error printer was unreadable spaghetti code before, and the change arguably makes it worse by simply making it longer and so even harder to understand.

I would propose the following structure:

```ocaml
let report_context ppf = function
  ...
in
let explain_mismatch expected 
let report_error ppf = function
  ...
in
fprintf ppf "@[<v>%a@;@[%a@]@]"
```

A piece of the puzzle is how to print either `It was expected` or `The 2nd parameter was expected` in a readable way. A proposal, sketched as a test toplevel session:

```ocaml
let explain_mismatch fmt =
  fmt ^^ " was expected to be %s,@ but it is %s."
;;

Printf.sprintf
  (explain_mismatch "It")
  "covariant" "contravariant";;

Printf.sprintf
  (explain_mismatch "The %d%s type parameter")
    2 "nd"
    "covariant" "contravariant";;
```

Reply:
In this specific case, it seems to me that closing the vertical box in the context message would be a simpler fix since it avoids the long-range effect of opening the word.

I agree that message could be refactorized, but that should be done in another PR.

In particular, I believe that playing word games in error messages with format string is a losing proposition in term of both readability and internationalization.  If we have two distinct formulations, it is better to have two distinct error messages.